# 4.5_Task Data Consistency Checks

## Contents
### Importing Libraries
### Importing Data
### Path to Storage
### Mixed-Type Data
### Missing Values
#### Finding Missing Values
#### Addressing Missing Values
### Duplicates
#### Finding Duplicates
#### Addressing Duplicates
## Task
##### Task 1
##### Task 2
##### Task 3
##### Task 4
##### Task 5
##### Task 6
##### Task 7
##### Task 8
##### Task 9
##### Task 10

### Importing Libraries

In [3]:
# importing libraries
import pandas as pd
import numpy as mp
import os

### Importing Data

In [5]:
# importing products and orders csv
df_prods = pd.read_csv(r'C:\Users\ihs10539\Achievement 4 Project\02_Data\02.01_Original Data\products.csv', index_col = False)
df_ords = pd.read_csv(r'C:\Users\ihs10539\Achievement 4 Project\02_Data\02.02_Prepared Data\4.4_Task_orders_wrangled.csv', index_col = False)

### Path to Storage

In [7]:
# Path to storage location
path = r'C:\Users\ihs10539\Achievement 4 Project'

# Mixed-Type Data

In [10]:
# Create a Dataframe
df_test = pd.DataFrame()

In [12]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [16]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
# change mix column to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

## Finding Missing Values

In [22]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [24]:
# viewing missing 16 values under product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [26]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Addressing Missing Values

In [29]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [31]:
df_prods_clean.shape

(49677, 5)

# Duplicates

## Finding Duplicates

In [35]:
# creates new dataframe to utilize duplicated() function to identify duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [37]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## Addressing Duplicates

In [40]:
# one last verification before dropping duplicates
df_prods_clean.shape

(49677, 5)

In [42]:
# fuction to drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [44]:
# verification duplicates were dropped
df_prods_clean_no_dups.shape

(49672, 5)

In [46]:
# exporting new prods csv to Prepared Data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02_Data', '02.02_Prepared Data', '4.5_products_checked.csv'))

# Task

### Task 1

In [50]:
# Completed during exercise
# did perform an import again to have the exported df_prods_clean_no_dups available
df_prods_task = pd.read_csv(r'C:\Users\ihs10539\Achievement 4 Project\02_Data\02.02_Prepared Data\4.5_products_checked.csv', index_col = False)

In [52]:
# verifying its correct just by checking the shape from the step right before export
df_prods_task.shape

(49672, 6)

### Task 2

In [55]:
# Creating new framework to use through task as to not directly work with original prod
df_ords_checked = df_ords

In [57]:
df_ords_checked.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

##### Surprisingly everything seems to be in order from what I can tell.  
##### My guess is due to us doing so much to this dataframe already in previous exercises that it is likely close to being tip top.

### Task 3

In [61]:
for col in df_ords_checked.columns.tolist():
    weird = (df_ords_checked[[col]].map(type) != df_ords_checked[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords_checked[weird]) > 0:
        print (col)

#### As you can see since there is no print out under the block above there doesn't appear to be any mixed type data in df_prods_task dataframe

### Task 4

In [65]:
# since I didn't locate any mixed data I thought I would at least find out what each column was aassigned
df_ords_checked.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   orders_day_of_week      int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


### Task 5

In [68]:
# Missing value check function
df_ords_checked.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### days_since_prior_order has 206209 missing values.
#### While this alarmed me at first it struck me that there is the exact same number of user_ids.
#### Meaning these missing numbers are from each of the users the day before their first order
#### ultimately this means we don't need to remove the missing values
#### NOTE: If desired more effort could possibly be put in to see if an additional column could be implemented to mark said missing values to hide them however it is not a must.

In [71]:
df_ords_checked

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3421083 rows x 8 columns]

### Task 6

#### Again as stated in Task 5 above the missing values corrispond to an understandable missing value for each of the user_ids. While it would be possible to potentially hide them through some extra effort it would likely be best practice to simply leave them and continue on.

### Task 7

In [76]:
# It's duplicate hunting season
df_ords_checked[df_ords_checked.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### Ok now I'm just getting nervous as there are no duplicates either. 
#### Chekcing the df_ords dataframe in a normal environment should there have been duplicates I would have validated whether they should be there or not. 
#### For example like in the reading an aisle or department will repeat but a user_id should not.
#### If it was something that should not repeat it would simply be removed using the .drop_duplicates() function

### Task 8

In [80]:
# Getting rid of Unamed: 0 since it does the same as the index
df_ords_checked = df_ords.drop('Unnamed: 0', axis=1)

In [82]:
df_ords_checked

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
1         2398795        1    prior             2                   3   
2          473747        1    prior             3                   3   
3         2254736        1    prior             4                   4   
4          431534        1    prior             5                   4   
...           ...      ...      ...           ...                 ...   
3421078   2266710   206209    prior            10                   5   
3421079   1854736   206209    prior            11                   4   
3421080    626363   206209    prior            12                   1   
3421081   2977660   206209    prior            13                   1   
3421082    272231   206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

#### Decided before performing the export in the next step I wanted to get rid of the Unamed: 0 column since it just mirrors what the index naturally does. 
#### Best guess is it was an index that pulled in from another data system so no reason to have two of them gunking up the works. 

### Task 9

In [88]:
# exporting csvs to Prepared Data folder
df_ords_checked.to_csv(os.path.join(path, '02_Data', '02.02_Prepared Data', '4.5_Task_orders_checked.csv'))
df_prods_task.to_csv(os.path.join(path, '02_Data', '02.02_Prepared Data', '4.5_Task_products_checked.csv'))

### Task 10

In [91]:
#saved and turned in. Thanks for all you do